iqps/upload/forms.py

```diff
@@ -1,17 +1,43 @@
+import json
 import logging
+import os
+import re
+from filelock import Timeout, FileLock
 from django import forms
 from django_select2.forms import ModelSelect2TagWidget, Select2Widget
 from captcha.fields import CaptchaField
+from django.core.exceptions import ValidationError 
 
 from data.models import Paper, Keyword
 from utils.timeutil import current_year
 
 LOG = logging.getLogger(__name__)
 
 
+BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+FILE_PATH = os.path.join(BASE_DIR, 'static/files/code_subjects.json')
+LOCK_PATH = os.path.join(BASE_DIR, 'static/files/code_subjects.json.lock')
+lock = FileLock(LOCK_PATH, timeout=3)
```

Comment:
let's keep a little higher timeout, say 5 seconds.

I would rather suggest try out adding a new field in the JSON and see how much time it takes. Accordingly, set the timeout to 10x of the observed time.

Reply:
i tested the time period between reading and writting to the same JSON file
it is in the order of milli seconds most of the times it is 5,6,7 milli seconds
max it was 10 milli seconds.
i stil feel 3 seconds should be reduced to 1 s or 2 second(considering lock will not take too much time)